arbitrum-docs/devs-how-tos/bridge-tokens/how-to-bridge-tokens-custom-gateway.mdx

```diff
@@ -6,6 +6,856 @@ reader_audience: developers who want to build on Ethereum/Arbitrum and bridge to
+
+Furthermore, if you plan on having permissionless registration of tokens in your gateway, your L1 gateway should also have a `registerCustomL2Token` method, similar to the one being used in Arbitrum’s [custom generic gateway](https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/ethereum/gateway/L1CustomGateway.sol#L146).
+
+On the other hand, the **L2 counterpart of the gateway**, must conform to the [ITokenGateway](https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/libraries/gateway/ITokenGateway.sol) interface. This means that it must have, at least:
+
+- A method `outboundTransfer`, to handle external calls, and forwarded calls from `L2GatewayRouter.outboundTransfer`.
+- A method `finalizeInboundTransfer`, to handle messages coming ONLY from L1’s gateway.
+- Two methods `calculateL2TokenAddress` and `getOutboundCalldata` to handle other bridging operations.
+- Methods to send cross-chain messages through ArbSys precompile. An example implementation can be found in  `sendTxToL1` on [L2ArbitrumMessenger](https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/arbitrum/L2ArbitrumMessenger.sol).
+
+### What about my custom tokens?
+
+If you are deploying custom gateways, you will probably want to support your custom tokens on L1 and L2 too. They also have several requirements they must comply with. You can find more information about it in [How to bridge tokens via Arbitrum’s custom generic gateway](/devs-how-tos/bridge-tokens/how-to-bridge-tokens-custom-generic.mdx).
+
+## Step 1: Create a gateway and deploy on L1
+
+:::caution This code is for testing purposes
+
+The code contained within the following sections is meant for testing purposes only and does not guarantee any level of security. It has not undergone any formal audit or security analysis, so it is not ready for production use. Please exercise caution and due diligence while using this code in any environment.
+
+:::
+
+We‘ll begin the process by creating and deploying on L1 our custom gateway. A good example of a custom gateway is [Arbitrum’s custom generic gateway](https://github.com/OffchainLabs/token-bridge-contracts/blob/main/contracts/tokenbridge/ethereum/gateway/L1CustomGateway.sol). It includes all methods required plus some more to support the wide variety of tokens that can be bridged through it.
+
+In this case, we’ll use a simpler approach. We’ll create a gateway that supports only one token and that has the ability to be disabled/enabled by the owner of the contract. It will also implement all necessary methods. To simplify the deployment process even further, we won’t worry about setting the addresses of the counterpart gateway and the custom tokens at deployment time. Instead, we will use a function `setTokenBridgeInformation` that will be called by the owner of the contract to initialize the gateway.
```

Comment:
```suggestion
In this case, we’ll use a simpler approach. We’ll create a gateway that supports only one token and has the ability to be disabled/enabled by the owner of the contract. It will also implement all necessary methods. To simplify the deployment process even further, we won’t worry about setting the addresses of the counterpart gateway and the custom tokens at deployment time. Instead, we will use a function `setTokenBridgeInformation` that will be called by the owner of the contract to initialize the gateway.
```